invisible_cities/reco/paolina_functions.py

```diff
@@ -176,31 +176,105 @@ def length(track: Graph) -> float:
+
+
+def hits_in_blob(track_graph : Graph, radius : float, extreme: Voxel) -> Sequence[BHit]:
+    """Returns the hits that belong to a blob."""
+    distances         = shortest_paths(track_graph)
+    dist_from_extreme = distances[extreme]
+    blob_pos          = blob_centre(extreme)
+    diag              = np.linalg.norm(extreme.size)
+
+    blob_hits = []
+    # First, consider only voxels at a certain distance from the end-point, along the track.
+    # We allow for 1 extra contiguity, because this distance is calculated between
+    # the centres of the voxels, and not the hits. In the second step we will refine the
+    # selection, using the euclidean distance between the blob position and the hits.
+    for v in track_graph.nodes():
+        voxel_distance = dist_from_extreme[v]
+        if voxel_distance < radius + diag:
+            for h in v.hits:
+                hit_distance = np.linalg.norm(blob_pos - h.pos)
+                if hit_distance < radius:
+                    blob_hits.append(h)
+
+    return blob_hits
+
+
 def blob_energies(track_graph : Graph, radius : float) -> Tuple[float, float]:
     """Return the energies around the extrema of the track."""
     distances = shortest_paths(track_graph)
-    a, b, _ = find_extrema_and_length(distances)
-    Ea = energy_within_radius(distances[a], radius)
-    Eb = energy_within_radius(distances[b], radius)
-    return (Ea, Eb) if Ea < Eb else (Eb, Ea)
+    a, b, _   = find_extrema_and_length(distances)
+    ha = hits_in_blob(track_graph, radius, a)
+    hb = hits_in_blob(track_graph, radius, b)
+
+    Ea = sum(h.E for h in ha)
+    Eb = sum(h.E for h in hb)
+
+    # Consider the case where voxels are built without associated hits
+    if len(ha) == 0 and len(hb) == 0 :
+        Ea = voxel_energy_within_radius(distances[a], radius)
+        Eb = voxel_energy_within_radius(distances[b], radius)
+
+    return (Eb, Ea) if Ea < Eb else (Ea, Eb)
+
+
+def blob_energies_and_hits(track_graph : Graph, radius : float) -> Tuple[float, float, Tuple[Tuple[float], Tuple[float]]]:
+    """Return the energies and the hits around the extrema of the track."""
+    distances = shortest_paths(track_graph)
+    a, b, _   = find_extrema_and_length(distances)
+    ha = hits_in_blob(track_graph, radius, a)
+    hb = hits_in_blob(track_graph, radius, b)
+
+    Ea = sum(h.E for h in ha)
+    Eb = sum(h.E for h in hb)
+
+    # Consider the case where voxels are built without associated hits
+    if len(ha) == 0 and len(hb) == 0 :
+        Ea = voxel_energy_within_radius(distances[a], radius)
+        Eb = voxel_energy_within_radius(distances[b], radius)
 
+    return (Eb, Ea, hb, ha) if Ea < Eb else (Ea, Eb, ha, hb)
 
-def compute_blobs(track_graph : Graph, radius : float) -> Tuple[Voxel, Voxel, List[Voxel], List[Voxel]]:
-    """Return the blobs (list of voxels) around the extrema of the track."""
+
+def blob_centres(track_graph : Graph, radius : float) -> Tuple[Tuple[float], Tuple[float]]:
```

Comment:
Something similar with this function. I would include the information of the blob centres into the same function that returns blob energies and hits.
Now there are 3 different functions that do basically the same computing, but they provide:
1st version: blob hits
2nd version: blob hits and energies
3rd version: blob centres
If someone needs for their analysis the energy and the centre of blobs, he would need to make the same computation twice.

Reply:
I have followed what was already implemented by @jacg. Maybe he can gives us some insight on why he decided to write several functions.

Reply:
I'm having problems contacting the LFS server, so I am unable to check out this branch, so it's not clear to me what the relationship is between `blob_centres` (the function we seem to be discussing) and any function I wrote. The only thing that I wrote in that file that contains `centre` in its name is `centre` in `voxelize_hits` If you are alluding to its similarity to `sizes` this is because

+ Each function does exactly one thing clearly and concisely, and is therefore trivial to implement and understand.
+ Each function has a trivial and easy to understand interface.
+ Each function has a name that clearly and concisely describes what it does.
+ These functions were originally written inside `voxelize_hits`, as implementation details of the latter: and were made as simple as possible, because they were written in order to make the code self-documenting, and were never intended for use by anyone outside of the function which they helped to implement.

But as I cannot check out this branch, I cannot be entirely sure that this is relevant to the question being asked.

Reply:
Hi Jacek. We're talking about the functions `blob_energies`, `blob_energies_and_hits` and `blob_centres`, which are all new functions. I decided not to write just one big function which returns 6 objects (blob energies, hits and centres), because it looked to me that it was clearer to separate functions, and I was following what you did with `find extrema`, `length` and `find extrema_and_length`, but it's possible that I completely missed your point there.

Reply:
I have written one function that returns all the quantities (blob energies, hits and centres) and three different functions that call this big one, to extract only what is needed.

Reply:
OK, I see what you are talking about now.

 The bodies of these three functions (`blob_energies`, `blob_energies_and_hits` and `blob_centres`) are *identical* in their first 9 lines, and only differ in the 1,1 or 3 lines that follow.

This copy'n'paste code reuse is, of course, very wrong, and *must* be removed.

Please write a single function which performs the first 9 common lines. Once this is done, there are a few obvious approaches to consider:

1. Have this function return the values it calculated: `((a, ha, Ea), (b, hb, Eb))`, where the `a`s refer to the higher-energy blob.
2. Parametrize this function with a function which extracts whatever combination of that information the user wants.

To be clear, here are some code samples that show precisely what I mean:

     # Do we already have a Blob type somewhere?
    Blob = namedtuple('Blob', 'voxel, hits, energy')
    
    # Option 1 (Just return the voxels, hits and energies calculated)
    def blob_whatever(track_graph : Graph, radius : float) -> Tuple[Blob, Blob]:
        distances = shortest_paths(track_graph)
        a, b, _   = find_extrema_and_length(distances)
        ha = hits_in_blob(track_graph, radius, a)
        hb = hits_in_blob(track_graph, radius, b)
    
        Ea = sum(h.E for h in ha)
        Eb = sum(h.E for h in hb)
    
        # Consider the case where voxels are built without associated hits
        if len(ha) == 0 and len(hb) == 0 :
            Ea = voxel_energy_within_radius(distances[a], radius)
            Eb = voxel_energy_within_radius(distances[b], radius)
    
        if Ea < Eb:
            a,ha,Ea, b,hb,Eb = b,hb,Eb, a,ha,Ea 
    
        return Blob(a,ha,Ea), Blob(b,hb,Eb)
    
    
    # Option 2 (pass in a function which determines what should be returned)
    def blob_whatever(track_graph : Graph, radius : float, xxx: Callable[[Blob, Blob], T]) -> T:
        distances = shortest_paths(track_graph)
        a, b, _   = find_extrema_and_length(distances)
        ha = hits_in_blob(track_graph, radius, a)
        hb = hits_in_blob(track_graph, radius, b)
    
        Ea = sum(h.E for h in ha)
        Eb = sum(h.E for h in hb)
    
        # Consider the case where voxels are built without associated hits
        if len(ha) == 0 and len(hb) == 0 :
            Ea = voxel_energy_within_radius(distances[a], radius)
            Eb = voxel_energy_within_radius(distances[b], radius)
    
        if Ea < Eb:
            a,ha,Ea, b,hb,Eb = b,hb,Eb, a,ha,Ea 
    
        return xxx(Blob(a,ha,Ea), Blob(b,hb,Eb))
    
    # With option 1, blob_energies could be implemented thus:
    def blob_energies(track_graph : Graph, radius : float) -> Tuple[float, float]:
        a,b = blob_whatever(track_graph, radius)
        return a.energy, b.energy
    
    # With option 2, blob_energies could be implemented thus:
    def blob_energies(track_graph : Graph, radius : float) -> Tuple[float, float]:
        return blob_whatever(track_graph, radius, lambda a,b: a.energy, b.energy)
    
    # Or, given that you are unlikely to want to extract different bits of
    # information off the two blobs, perhaps the following variation on
    # the second theme would be better
    
    def blob_whatever(track_graph : Graph, radius : float, xxx: Callable[[Blob], T]) -> Tuple[T,T]:
        distances = shortest_paths(track_graph)
        a, b, _   = find_extrema_and_length(distances)
        ha = hits_in_blob(track_graph, radius, a)
        hb = hits_in_blob(track_graph, radius, b)
    
        Ea = sum(h.E for h in ha)
        Eb = sum(h.E for h in hb)
    
        # Consider the case where voxels are built without associated hits
        if len(ha) == 0 and len(hb) == 0 :
            Ea = voxel_energy_within_radius(distances[a], radius)
            Eb = voxel_energy_within_radius(distances[b], radius)
    
        if Ea < Eb:
            a,ha,Ea, b,hb,Eb = b,hb,Eb, a,ha,Ea 
    
        return xxx(Blob(a,ha,Ea)), xxx(Blob(b,hb,Eb))
    
    # in which case blob_energies could look like this:
    def blob_energies(track_graph : Graph, radius : float) -> Tuple[float, float]:
        return blob_whatever(track_graph, radius, lambda blob: blob.energy)
    
I have a strong preference for option 2, but if too many people on the team find it more confusing than option 1, then perhaps you should go with option 1. (Or one could say that if too many people are confused by option 2, then we should educate them until they find it me most natural thing to do!)

Whatever you choose, the copy'n'paste code reuse has to go as it violates the **DRY** principle: **D**o not **R**epeat **Y**ourself.

Also, don't forget that

    lambda blob: blob.energy
    lambda blob: blob.energy, blob.voxel

can be written as 

    attrgetter('energy')
    attrgetter('energy', 'voxel')

Reply:
Thanks Jacek! I already changed the code following the DRY principle, without passing through the `Blob`class, which we have, by the way. I just wrote three different functions that call the same one, so the code is not repeated. Do you think it is fine?

Reply:
This is much better.

I'm not too enthusiastic about `blob_energies_and_hits` returning 4 essentially anonymous values, but it's not obvious how to fix that without complicating things too much.

The end of `blob_energies_hits_and_centres` could benefit from having the two alternative return values being aligned one above the other, to facilitate seeing what the difference is. The cleanest way to spell it is probably:

    if Ea < Eb: return (Eb, Ea, hb, ha, cb, ca) 
    else      : return (Ea, Eb, ha, hb, ca, cb)

Actually, the highest-energy-blob-first condition is probably expressed more clearly like this:

    if Eb > Ea: return (Eb, Ea, hb, ha, cb, ca) 
    else      : return (Ea, Eb, ha, hb, ca, cb)